src/core/server/application/application_context_provider.test.ts

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { BehaviorSubject } from 'rxjs';
+
+import { mockCoreContext } from '../core_context.mock';
+import { httpServiceMock } from '../http/http_service.mock';
+import { httpServerMock } from '../http/http_server.mocks';
+import { pluginServiceMock } from '../plugins/plugins_service.mock';
+import { uiSettingsServiceMock } from '../ui_settings/ui_settings_service.mock';
+import { ApplicationContextProvider } from './application_context_provider';
+
+const HEADERS_SNAPSHOT = `
+  Object {
+    "content-security-policy": "script-src 'unsafe-eval' 'self'; worker-src blob:; child-src blob:; style-src 'unsafe-inline' 'self'",
+  }
+`;
+
+describe('ApplicationContextProvider', () => {
+  let context: ReturnType<typeof mockCoreContext.create>;
+  let http: ReturnType<typeof httpServiceMock.createSetupContract>;
+  let plugins: ReturnType<typeof pluginServiceMock.createSetupContract>;
+  let request: ReturnType<typeof httpServerMock.createKibanaRequest>;
+  let response: ReturnType<typeof httpServerMock.createResponseFactory>;
+  let uiSettings: ReturnType<typeof uiSettingsServiceMock.createClient>;
+  let provider: ApplicationContextProvider;
+
+  beforeEach(async () => {
+    context = mockCoreContext.create();
+    http = httpServiceMock.createSetupContract();
+    plugins = pluginServiceMock.createSetupContract();
+    request = httpServerMock.createKibanaRequest();
+    response = httpServerMock.createResponseFactory();
+    uiSettings = uiSettingsServiceMock.createClient();
+    provider = new ApplicationContextProvider({
+      env: context.env,
+      http,
+      plugins,
+      request,
+      response,
+      uiSettings,
+    });
+
+    uiSettings.getUserProvided.mockResolvedValue({
+      providedBy: {
+        isOverridden: true,
+        value: 'user',
+      },
+    });
+  });
+
+  describe('#render()', () => {
+    it('renders "core" application by default', async () => {
+      await provider.render();
+      expect(response.ok).toHaveBeenCalled();
+
+      const { headers, body } = response.ok.mock.calls[0][0]!;
+
+      expect(headers).toMatchInlineSnapshot(HEADERS_SNAPSHOT);
+      expect(body).toMatch('app:core');
```

Comment:
It'd be great if we could de-couple these tests from the HTML template. Maybe there could be another level of abstraction (probably just a function) that sits between the calculation of the variables for the template and the actual rendering of the template. This would allow us to test the calculated variables more precisely than doing simple string matches against the rendered HTML.

In the future this abstraction would probably be the "RenderingService", but for now we can do something simple here.

Reply:
Sounds good!